tchannel/tornado/connection.py

```diff
@@ -471,25 +470,21 @@ def serve(self, handler):
                 # TODO Send error frame back
                 logging.exception("Failed to process %s", repr(message))
 
-    def send_error(self, code, description, message_id):
+    def send_error(self, error):
         """Convenience method for writing Error frames up the wire.
 
-        :param code:
-            Error code
-        :param description:
-            Error description
-        :param message_id:
-            Message in response to which this error is being sent
+        :param error:
+            TChannel Error.
         """
-        if code not in ErrorMessage.ERROR_CODES.keys():
-            raise FatalProtocolError(code)
-
-        return self._write(
-            ErrorMessage(
-                code=code,
-                description=description,
-                id=message_id,
-            ),
+
+        error_message = build_raw_error_message(error)
+        write_future = self._write(error_message)
+        tornado.ioloop.IOLoop.current().add_future(
```

Comment:
So there's an important difference between `add_future` like this and `add_done_callback`. `add_future` schedules things such that another coroutine can run between our future finishing and an event hook firing. `add_done_callback` will execute the hook immediately after the future finishes.

Since event hooks tend to be concerned about things like timing, logging, etc., I think we actually *do* want to use `add_done_callback` with our event hooks. @abhinav would you agree with that?